system/lib/compiler-rt/stack_ops.s

```diff
@@ -0,0 +1,32 @@
+.globl stackSave
+.globl stackRestore
+.globl stackAlloc
+
+.globaltype __stack_pointer, i32
+
+stackSave:
+  .functype stackSave() -> (i32)
+  global.get __stack_pointer
+  end_function
+
+stackRestore:
+  .functype stackRestore(i32) -> ()
+  local.get 0
+  global.set __stack_pointer
+  end_function
+
+stackAlloc:
+  .functype stackAlloc(i32) -> (i32)
+  .local i32, i32
+  global.get __stack_pointer
+  # Get arg 0 -> number of bytes to allocate
+  local.get 0
+  # Stack growns down.  Subtrace arg 0 from __stack_pointer
+  i32.sub
+  # Align result by anding with ~15
+  i32.const 0xfffffff0
+  i32.and
+  local.tee 1
+  global.set __stack_pointer
+  local.get 1
+  end_function
```

Comment:
Fascinating approach..!

The current codegen in binaryen can add stack check assertions to this. How would that work here? Would we need to have multiple versions of `.s`, with and without?

Reply:
Assemblers usually support the C preprocessor (traditionally I think these use `.S` extensions instead of `.s`); could we have inline ifdefs?

Reply:
Ah!  I'll need to take a look at that.

Reply:
Actually I don't think its necessary.

We already have the `StackLimitEnforcer` which finds all the assignments to stack pointer global.   This one will be no different to any other that is generated by llvm, right?

Reply:
That sounds plausible, but it implies the binaryen code was unnecessary (we could have relied on StackLimitEnforcer more there). We should verify this, since it's not clear to me that our tests are sufficient, as we've noticed recently in https://github.com/WebAssembly/binaryen/issues/2850

Reply:
> lgtm with the stack checks confirmed to be identical to before this PR.

Do you want me to add a test that verifies that `StackLimitEnforcer` works specifically when growing the stack with stackAlloc?